stdlib/2/urllib2.pyi

```diff
@@ -68,50 +77,52 @@ class HTTPErrorProcessor(BaseHandler):
     def http_response(self, request, response): ...
 
 class HTTPDefaultErrorHandler(BaseHandler):
-    def http_error_default(self, req, fp, code, msg, hdrs): ...
+    def http_error_default(self, req: Request, fp: IO, code: int, msg: str, hdrs: Message): ...
 
 class HTTPRedirectHandler(BaseHandler):
     max_repeats = ...  # type: int
     max_redirections = ...  # type: int
-    def redirect_request(self, req, fp, code, msg, headers, newurl): ...
-    def http_error_301(self, req, fp, code, msg, headers): ...
-    def http_error_302(self, req, fp, code, msg, headers): ...
-    def http_error_303(self, req, fp, code, msg, headers): ...
-    def http_error_307(self, req, fp, code, msg, headers): ...
+    def redirect_request(self, req: Request, fp: IO, code: int, msg: str, headers: Message, newurl): ...
+    def http_error_301(self, req: Request, fp: IO, code: int, msg: str, headers: Message): ...
+    def http_error_302(self, req: Request, fp: IO, code: int, msg: str, headers: Message): ...
+    def http_error_303(self, req: Request, fp: IO, code: int, msg: str, headers: Message): ...
+    def http_error_307(self, req: Request, fp: IO, code: int, msg: str, headers: Message): ...
     inf_msg = ...  # type: str
 
 
 class ProxyHandler(BaseHandler):
     def __init__(self, proxies: Optional[Any] = ...): ...
-    def proxy_open(self, req, proxy, type): ...
+    def proxy_open(self, req: Request, proxy, type): ...
 
 class HTTPPasswordMgr:
     def __init__(self) -> None: ...
-    def add_password(self, realm: _string, uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
-    def find_user_password(self, realm: _string, authuri: _string) -> Tuple[Any, Any]: ...
+    def add_password(self, realm: Optional[_string], uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
+    def find_user_password(self, realm: Optional[_string], authuri: _string) -> Tuple[Any, Any]: ...
```

Comment:
Is `realm` really supposed to be `None`? While the implementation would work with `None` values, lines like the following make me assume that this is not really a supported use case:

https://github.com/python/cpython/blob/34b7c438b8dc0a1e7e23c9b2d7ce7f8a7c31b4f4/Lib/urllib2.py#L787

Also applies to two more instances below.

Reply:
Quoting the commit message of my own b7744a9c872bf75688e4fe5392cfc71a8eeea0f6:
> Also realm=None is allowed, which is required when
> HTTPPasswordMgrWithDefaultRealm is used:
> <https://docs.python.org/2/howto/urllib2.html#id6>:
> > Frequently one doesn’t care what the realm is.
> ...
> > We indicate this by providing None as the realm argument to the
> > add_password method.

See <https://github.com/python/cpython/blob/34b7c438b8dc0a1e7e23c9b2d7ce7f8a7c31b4f4/Lib/urllib2.py#L849> where `find_user_password(self, realm, authuri)` is called with `realm=None`